### PR TITLE
Add functionality to run from_actors mode with more than one TF

### PR DESCRIPTION
**Pull request by roskamsh**
When running interactively, all TargeneCore tests pass with no errors.

### CODE REVIEW

File: src/tmle_inputs/from_actors.jl [hunk 1]
```diff
@@ -44,7 +62,7 @@ all_variants(bqtls::DataFrame, transactors::Vector{DataFrame}) = Set(vcat(bqtls.
 
 
 read_snps_from_csv(path::Nothing) = nothing
-read_snps_from_csv(path::String) = unique(CSV.read(path, DataFrame; select=[:ID, :CHR]), :ID)
+read_snps_from_csv(path::String) = unique(CSV.read(path, DataFrame; select=[:ID, :CHR, :TF]), [:ID, :TF])
```
Comment (olivierlabayle) [inline src/tmle_inputs/from_actors.jl:65]: So this is a breaking change, the TF column is not an option.

File: src/tmle_inputs/from_actors.jl [hunk 2]
```diff
@@ -12,6 +12,24 @@ function treatments_from_actors(bqtl_file, env_file, trans_actors_prefix)
 end
 
 
+tf_error() = throw(ArgumentError("No TF specified to filter SNPs on"))
+missing_snps_error() = throw(ArgumentError("No SNPs provided to filter on"))
+
+filter_snps_by_tf(df::Nothing, tf_name::Nothing) = nothing
+filter_snps_by_tf(df::DataFrame, tf_name::Nothing) = tf_error()
```
Comment (olivierlabayle) [inline src/tmle_inputs/from_actors.jl:19]: I am not sure what these (or the above) cover?

File: test/confounders.jl [hunk 3]
```diff
@@ -56,7 +56,6 @@ end
     
     clean(parsed_args)
 
-    # No qc file provided
```
Comment (olivierlabayle) [inline test/confounders.jl:59]: why is that comment removed?

File: src/tmle_inputs/from_actors.jl [hunk 4]
```diff
@@ -208,11 +233,18 @@ function tmle_inputs_from_actors(parsed_args)
...
-    TargeneCore.parameters_from_actors(
-        bqtls, transactors, data, variables, orders, outprefix; 
+
+    # Loop through each TF present in bqtls file 
+    tfs = unique(bqtls.TF)
+    for tf in tfs
+        bqtls_tf = TargeneCore.filter_bqtls_by_tf(bqtls, tf)
+        transactors_tf = TargeneCore.filter_transactors_by_tf(transactors, tf)
+        TargeneCore.parameters_from_actors(
+        bqtls_tf, transactors_tf, data, variables, orders, outprefix, tf;
```
Comment (olivierlabayle) [inline src/tmle_inputs/from_actors.jl:243]: I think it will make the code simpler to make a new `outprefix` for each TF instead of passing a new argument what do you think?

File: src/tmle_inputs/from_actors.jl [hunk 5]
```diff
@@ -148,22 +166,29 @@ function get_variables(pcs, traits, extraW, extraC, extraT)
...
+function parameters_from_actors(bqtls, transactors, data, variables, orders, outprefix, tf_name; positivity_constraint=0., batch_size=nothing)
     parameters = TMLE.Parameter[]
     batch_id = 1
     extraT_df = variables.extra_treatments isa Nothing ? nothing : DataFrame(ID=variables.extra_treatments)
     # For each interaction order generate parameter files
     for order in orders
         # First generate the `T` section
         treatment_combinations = TargeneCore.combine_by_bqtl(bqtls, transactors, extraT_df, order)
+        # If there are duplicates here, remove them
+        treatment_combinations = unique(treatment_combinations)
```
Comment (olivierlabayle) [inline src/tmle_inputs/from_actors.jl:178]: Is the previous function generating duplicates?

File: src/tmle_inputs/tmle_inputs.jl [hunk 6]
```diff
@@ -10,13 +16,13 @@ The SAMPLE_ID column should be read as a String.
 """
 read_data(filepath) = CSV.read(filepath, DataFrame, types=Dict(:SAMPLE_ID => String))
 
-function write_tmle_inputs(outprefix, final_dataset, parameters; batch_size=nothing)
+function write_tmle_inputs(outprefix, final_dataset, parameters; tf_name=nothing, batch_size=nothing)
```
Comment (olivierlabayle) [inline src/tmle_inputs/tmle_inputs.jl:19]: why did you need to change that function? I don't think it is called in the from_actors mode

**Review by olivierlabayle — CHANGES_REQUESTED** [review]
Thank you for taking over this Breeshey, I've started the review. I think there is one major point and one less important but still nice to address before I move on with checking correctness and the test. I've also added some comments regarding code style, some superfluous functions and some changes that I think need to be undone.

1. Unless I am missing something, the change is breaking at the user level because now the TF column is mandatory. I think it would be best to keep it optional because users only looking at a single transcription factor will not necessarily find useful to define this column. 
2. I think the change in the name of the ouput would be more easily read if the outprefix was changed instead.

In any case, because of the output name change this PR will be breaking at the pipeline's level and may require some changes in the targene-pipeline itself (this is entirely fine!).

File: src/tmle_inputs/from_actors.jl [hunk 7]
```diff
@@ -12,6 +12,24 @@ function treatments_from_actors(bqtl_file, env_file, trans_actors_prefix)
...
+filter_snps_by_tf(df::Nothing, tf_name::AbstractString) = missing_snps_error()
+function filter_snps_by_tf(df::DataFrame, tf_name::AbstractString)
+    return filter(row -> row.TF == tf_name, df)
+end
+
+function filter_bqtls_by_tf(bqtl_df::DataFrame, tf_name::AbstractString)
+    return filter_snps_by_tf(bqtl_df, tf_name)
+end
+
+function filter_transactors_by_tf(trans_actors::Union{Vector{DataFrame}}, tf_name::AbstractString)
```
[the remark this answers is not in the document]
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl:29]: Will this require passing in the bqtls variable as a vector? ie. when you pass in the bqtls variable into the parameters_from_actors() function, will you need to transform it to [bqtls]? Because currently it is just a DataFrame.

File: src/tmle_inputs/from_actors.jl [hunk 8]
```diff
@@ -12,6 +12,24 @@ function treatments_from_actors(bqtl_file, env_file, trans_actors_prefix)
 end
 
 
+tf_error() = throw(ArgumentError("No TF specified to filter SNPs on"))
+missing_snps_error() = throw(ArgumentError("No SNPs provided to filter on"))
+
+filter_snps_by_tf(df::Nothing, tf_name::Nothing) = nothing
+filter_snps_by_tf(df::DataFrame, tf_name::Nothing) = tf_error()
+filter_snps_by_tf(df::Nothing, tf_name::AbstractString) = missing_snps_error()
+function filter_snps_by_tf(df::DataFrame, tf_name::AbstractString)
```
[the remark this answers is not in the document]
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl:21]: Can definitely do this!

File: src/tmle_inputs/from_actors.jl (hunk 1 above)
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl:65]: How would you propose making this optional? Just read the CSV file in without selecting columns, and then check whether TF is in names(df)?

File: src/tmle_inputs/from_actors.jl (hunk 2 above)
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl:19]: I was just trying to cover edge cases in case the function didn't get passed the correct arguments, but I guess this may not happen?

File: test/confounders.jl (hunk 3 above)
Comment (roskamsh) [reply test/confounders.jl:59]: Unsure, it may have happened when I was resolving merge conflicts. I can add this back in.

File: src/tmle_inputs/from_actors.jl (hunk 5 above)
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl:178]: This can happen if different types of trans-actors for the same TF share the same RSID.

File: src/tmle_inputs/tmle_inputs.jl (hunk 6 above)
Comment (roskamsh) [reply src/tmle_inputs/tmle_inputs.jl:19]: I updated this because this function write_tmle_inputs() is also called from the function tmle_inputs_from_param_files(), and I needed to set a default tf_name given the way I organized it. I think if we make TF optional as mentioned above this can be reverted.

File: src/tmle_inputs/from_actors.jl (hunk 2 above)
Comment (olivierlabayle) [reply src/tmle_inputs/from_actors.jl:19]: yes I don't see how this would be happening at the moment

File: src/tmle_inputs/from_actors.jl (hunk 5 above)
Comment (olivierlabayle) [reply src/tmle_inputs/from_actors.jl:178]: Hum, let's discuss about that

File: src/tmle_inputs/from_actors.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl]: addressed: e4c2e131323f5032ff8a3be17776638ec0b1ba05

File: src/tmle_inputs/from_actors.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl]: addressed: e4c2e131323f5032ff8a3be17776638ec0b1ba05

File: src/tmle_inputs/from_actors.jl (hunk 1 above)
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl]: addressed: e4c2e131323f5032ff8a3be17776638ec0b1ba05

File: src/tmle_inputs/from_actors.jl (hunk 2 above)
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl]: addressed: e4c2e131323f5032ff8a3be17776638ec0b1ba05 

However - now there is a version of this function where tf_name::Nothing is passed in and used. This happens when there is no TF in the column names of the input bqtls.csv or trans_actors.csv files.

File: test/confounders.jl (hunk 3 above)
Comment (roskamsh) [reply test/confounders.jl:59]: added back in.

File: src/tmle_inputs/from_actors.jl (hunk 4 above)
Comment (roskamsh) [reply src/tmle_inputs/from_actors.jl]: outprefix has been updated to include the TF name: e4c2e131323f5032ff8a3be17776638ec0b1ba05 and e4c2e131323f5032ff8a3be17776638ec0b1ba05

File: src/tmle_inputs/tmle_inputs.jl (hunk 6 above)
Comment (roskamsh) [reply src/tmle_inputs/tmle_inputs.jl:13]: reverted back to original: 219f947a0878e0aacd61a8f7e72730c31d356660

**Review by olivierlabayle — APPROVED** [review]
Thank you Breeshey, this looks good to me! Good usage of multiple dispatch in the `filter_snps_by_tf` function. The missed covered line might be because they are smart enough to realise that the condition is never met in the tests. I've added [this issue](https://github.com/TARGENE/TargeneCore.jl/issues/186) for future reference. But nothing to be concerned about for now! Feel free to merge, thanks!
